### DISCUSSION TITLE
REST API for authentication?

### DISCUSSION THREAD

**Discussion by wez** [lldap/lldap#404] [Q&A] [answered]
This is a bit of a weird request, but here's how I got here:

* Home Assistant has a rudimentary auth provider system that can shell out to a script to auth and return the display name of the user
* They suggest using https://github.com/bob1de/ldap-auth-sh to perform LDAP auth
* But don't ship any ldap support in the docker image https://github.com/home-assistant/supervisor/issues/940

If I could `curl https://lldap/authenticate?user=$user&pass=$pass` and get something like `{"cn": "Full Name"}` back as a response, I could satisfy my desire to authenticate against lldap.

Is this something that we can do with lldap today, or if not, is it something you'd be open to supporting?

**Comment by nitnelave** [comment, marked as answer]
We have a simple login API usable by `curl`:
`curl https://lldap/simple/login -H 'Content-Type: application/json' -d '{"username":"my_login","password":"my_password"}'`
That'll give you a JWT as response, which you can include in a follow-up query to the graphql endpoint to get your full name.

Does that work for you?

**Reply by wez** [reply]
It sounds like it might, thanks!  I wrote up a thing using authelia (https://github.com/authelia/authelia/discussions/4681) that also needs to do two queries; I'll also try this out more directly against my lldap today.

Are there docs for the graphql query?

**Reply by nitnelave** [reply]
I think a mix of a graphql/curl tutorial, the schema reference (https://github.com/nitnelave/lldap/blob/main/schema.graphql) and the interactive endpoint at https://lldap/api/graphql/playground will help you find your way around.
Note that in the playground, you should set the header:
`{ "Authorization":"Bearer eyTko...." }` with your JWT token

In your case, you probably want something like:
```graphql
{
  user(userId:"admin") {
    displayName, 
  }
}
```

**Reply by wez** [reply]
Just following up on this: I was able to get something like this working, but wanted to note that the URL for simple auth was `https://lldap/auth/simple/login`; `/simple/login` 404s and it took me a few minutes to track down the `/auth` prefix from https://github.com/nitnelave/lldap/blob/09a0522e2df0ef31cdb2b2fc98efca82ede18e99/server/src/infra/tcp_server.rs#L84

Thanks!
